syntax = "proto3";

package model_trainer;

import "errors.proto";

service ModelTrainer {
  rpc Train (TrainRequest) returns  (TrainResponse);
  rpc ListAvailableModels (ListAvailableModelsRequest) returns (ListAvailableModelsResponse);
  rpc GetPrediction (GetPredictionRequest) returns (GetPredictionResponse);
}

message TrainRequest {
    string modelClass = 1;
    map<string, string> hyperParams = 2;
    bytes features = 3;  // First file
    bytes labels = 4;    // Second file
    string token = 5;
}

message TrainResponse {
 string modelClass = 1;
 optional errors.Error error = 2;
}

message ListAvailableModelsRequest {
  optional string token = 1;
}

message ListAvailableModelsResponse {
  repeated string modelClasses = 1;
  optional errors.Error error = 2;
}

message GetPredictionRequest {
  string modelClass = 1;
  optional string token = 2;
  bytes fileData = 3; // New field to include file data
}

message GetPredictionResponse {
  repeated float prediction = 1;
  optional errors.Error error = 2;
}